### PR TITLE
added DataApproximator+N extension

### PR DESCRIPTION
**Pull request by 666tos**
This implementation does not require tolerance for approximation and takes number of resulting points as parameter instead.
Uses modified Douglas-Peucker algorithm described here - http://psimpl.sourceforge.net/douglas-peucker.html

### CODE REVIEW

**Comment by liuxuan30** [comment]
wow, @danielgindi you should look at this :)

File: Source/Charts/Filters/DataApproximator+N.swift [hunk 1]
```diff
@@ -0,0 +1,152 @@
...
+        let dy = linePoint2.y - linePoint1.y
+        
+        let dividend = fabs(dy * self.x - dx * self.y - linePoint1.x * linePoint2.y + linePoint2.x * linePoint1.y)
+        let divisor = sqrt(dx * dx + dy * dy)
+        
+        return dividend/divisor
+    }
+}
+
+fileprivate class LineAlt {
```
Comment (jjatie) [inline Source/Charts/Filters/DataApproximator+N.swift]: `private`

File: Source/Charts/Filters/DataApproximator+N.swift [hunk 2]
```diff
@@ -0,0 +1,152 @@
...
+
+extension CGPoint {
+    fileprivate func distanceToLine(from linePoint1: CGPoint, to linePoint2: CGPoint) -> CGFloat {
+        let dx = linePoint2.x - linePoint1.x
+        let dy = linePoint2.y - linePoint1.y
+        
+        let dividend = fabs(dy * self.x - dx * self.y - linePoint1.x * linePoint2.y + linePoint2.x * linePoint1.y)
+        let divisor = sqrt(dx * dx + dy * dy)
+        
+        return dividend/divisor
```
Comment (jjatie) [inline Source/Charts/Filters/DataApproximator+N.swift]: `dividend / divisor` to keep consistent spacing

File: Source/Charts/Filters/DataApproximator+N.swift [hunk 3]
```diff
@@ -0,0 +1,152 @@
...
+        self.end = end
+        
+        let startPoint = points[start]
+        let endPoint = points[end]
+        
+        guard (end > start + 1) else {
+            return
+        }
+        
+        for i in start+1..<end {
```
Comment (jjatie) [inline Source/Charts/Filters/DataApproximator+N.swift]: `start + 1..<end`, though throughout Charts it is sometimes written as `start + 1 ..< end`

File: Source/Charts/Filters/DataApproximator+N.swift [hunk 4]
```diff
@@ -0,0 +1,152 @@
...
+    static func <(lhs: LineAlt, rhs: LineAlt) -> Bool {
+        return lhs.distance < rhs.distance
+    }
+}
+
+
+extension DataApproximator {
+    /// uses the douglas peuker algorithm to reduce the given arraylist of entries to given number of points
+    /// More algorithm details here - http://psimpl.sourceforge.net/douglas-peucker.html
+    @objc open class func reduceWithDouglasPeukerN(_ points: [CGPoint], resultCount: Int) -> [CGPoint]
```
Comment (jjatie) [inline Source/Charts/Filters/DataApproximator+N.swift]: The algorithm shouldn't be overridden. `@objc public static func reduceWithDouglasPeukerN(_ points: [CGPoint], resultCount: Int) -> [CGPoint]`

File: Source/Charts/Filters/DataApproximator+N.swift [hunk 5]
```diff
@@ -0,0 +1,152 @@
...
+        // create a new array with series, only take the kept ones
+        let reducedEntries = points.enumerated().flatMap { (index: Int, point: CGPoint) -> CGPoint? in
+            return keep[index] ? point : nil
+        }
+        
+        return reducedEntries
+    }
+    
+    // Keeps array sorted
+    private class func insertLine(_ line: LineAlt, into array: inout [LineAlt]) {
```
Comment (jjatie) [inline Source/Charts/Filters/DataApproximator+N.swift]: `private static func`

File: Source/Charts/Filters/DataApproximator+N.swift [hunk 6]
```diff
@@ -0,0 +1,152 @@
...
+        let insertionIndex = self.insertionIndex(for: line, into: &array)
+        array.insert(line, at: insertionIndex)
+    }
+    
+    private class func insertionIndex(for line: LineAlt, into array: inout [LineAlt]) -> Int {
```
Comment (jjatie) [inline Source/Charts/Filters/DataApproximator+N.swift]: `private static func`

**Comment by jjatie** [comment]
Awesome!

File: Source/Charts/Filters/DataApproximator+N.swift (hunk 4 above)
Comment (666tos) [reply Source/Charts/Filters/DataApproximator+N.swift]: Agree, but in current code there is 
'@objc open class func reduceWithDouglasPeuker(_ points: [CGPoint], tolerance: CGFloat) -> [CGPoint]'

Comment (jjatie) [reply Source/Charts/Filters/DataApproximator+N.swift]: That should be changed to. This is fine then. I will change to static in a separate PR.

**Comment by 666tos** [comment]
Fixed PR notes

**Pull request merged**